### PR TITLE
Hashing Data

### PR DESCRIPTION
**Pull request by jrans**
## THIS PR

Necessary for password, secret answers etc. Function can be used at layer most appropriate
- Function to take some data and hash asynchronously according to field settings in config.
- Fields can be given password or hash prop set as true to be told to hash.
- New object given back, old not mutated.
- Salt round defaults 10 but customisable with config object

Related: #52

### CODE REVIEW

**Comment by jrans** [comment]
CI needs fixing but code itself passes on my machine and so can start your review while I fix, will remove WIP when done

**Comment by jrans** [comment]
DONE Thanks to https://github.com/ncb000gt/node.bcrypt.js/issues/366

File: lib/hash.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+function successiveHash (data, hashed, hashQueue, cb) {
+  var toBeHashed = hashQueue.pop();
+
+  if (typeof toBeHashed === 'undefined') {
+    return cb(null, hashed);
+  }
+
+  return bcrypt.hash(data[toBeHashed], 10, function (err, hash) {
```
Comment (nelsonic) [inline lib/hash.js]: Ok. so we are _hard-coding_ the `bcrypt` hashing "_cycles_" in the library.
also, how are we going to _use_ this hashed data?
if we attempt to `bcrypt.compare` we need to have the plaintext 
see: https://github.com/ncb000gt/node.bcrypt.js/#to-check-a-password

**Comment by jrans** [comment]
@nelsonic Yes salt number should be configurable. Will change. 

This function does indeed only serve purpose for saving data that needs to be hashed. 

Actually using that hash to authenticate a login must be done separately but will be done later with #53

**Comment by nelsonic** [comment]
@jrans what I'm saying is that _given_ that the _purpose_ of this PR is to hash ("_sensitive_") data and the _only_ situation where we _need_ to `hash` (_i.e. one-way "encrypt"_) data is for passwords which we need the `plaintext` for in order to use `bcrypt.compare`  (_unless there is another MVP usecase I'm not aware of_) the functionality becomes [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) so PR _could_ be _closed_ 

image

> _the feature/idea should be documented in the issue so we can come back to it if/when we have a clear use-case_

**Comment by jrans** [comment]
@nelsonic Maybe you're missing what I'm proposing but this is function is used on insertion of a new user and password into the database. If the field is meant to be a password it will hash the data if not then it will leave it as is. 

Yes the only field we need to hash is `"password"`, but what if we had two password fields? What if we wanted to store secret answers? Although this is maybe above and beyond MVP its functionality that can be used in future. 

Two options:
- I'm happy to strip out the salt stuff and provide a function that just hashes a single predefined "password" field in the db before saving a new user.
  - Use this function as is at the db insert layer and set password field to have prop password true which will do the same.

Realise the latter is one more step for the user and we're probably assuming email, password and username necessary for all user data so this could be done by us anyway when applying default requirements on the config.

**Comment by jrans** [comment]
@nelsonic ok closing

**Pull request closed without merging**